symengine/ntheory.cpp

```diff
@@ -1447,6 +1524,72 @@ void powermod_list(std::vector<RCP<const Integer>> &pows, const RCP<const Intege
     }
 }
 
+std::vector<integer_class> quadratic_residues(const Integer &a)
+{
+    /*
+        Returns the list of quadratic residues.
+        Example
+        ========
+        >>> quadratic_residues(7)
+        [0, 1, 2, 4]
+    */
+
+    if (a.as_mpz() < 1) {
+        throw std::runtime_error("quadratic_residues: Input must be > 0");
+    }
+
+    std::vector<integer_class> residue;
+    for(integer_class i = integer_class(0); i <= a.as_int()/2; i++)
+    {
+        residue.push_back((i*i) % a.as_int());
+    }
+
+    sort(residue.begin(), residue.end());
+    residue.erase(unique(residue.begin(), residue.end()), residue.end());
+
+    return residue;
+
+}
+
+bool is_quad_residue(const Integer &a, const Integer &p)
+{
+    /*
+    Returns true if ``a`` (mod ``p``) is in the set of squares mod ``p``,
+    i.e a % p in set([i**2 % p for i in range(p)]). If ``p`` is an odd but
+    not prime, an iterative method is used to make the determination.
+    */
+
+    integer_class p2 = p.as_mpz();
+    if (p2 < 1)
+        throw std::runtime_error("is_quad_residue: Second parameter must be > 0");
+
+    integer_class a_final = a.as_mpz();
+    if (a.as_mpz() >= p2 || a.as_mpz() < 0)
+        mp_fdiv_r(a_final, a.as_mpz(), p2);
+    if (a_final < 2 || p2 < 3)
```

Comment:
Its not necessary but since no computation was required for them I decided to add this statement and return `true` straight away.